From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hpfxd <me@hpfxd.com>
Date: Tue, 26 Oct 2021 20:51:53 -0400
Subject: [PATCH] PandaSpigot version checking


diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
index e91e1d8c1215d19a78611d3c8302ed9634bfbda5..62f36e8f9d033c82318141e4edc0636cfc58613a 100644
--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -18,15 +18,20 @@ import com.google.common.io.Resources;
 import java.io.BufferedReader;
 import java.io.IOException;
 import java.net.URL;
-import java.net.URLEncoder;
 import java.util.HashSet;
 import java.util.Set;
 import java.util.concurrent.locks.ReentrantLock;
-import java.util.logging.Level;
-import java.util.logging.Logger;
-import org.json.simple.JSONObject;
-import org.json.simple.parser.JSONParser;
-import org.json.simple.parser.ParseException;
+
+// PandaSpigot start
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonSyntaxException;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.util.stream.StreamSupport;
+import com.google.gson.JsonElement;
+// PandaSpigot end
 
 public class VersionCommand extends BukkitCommand {
     public VersionCommand(String name) {
@@ -166,7 +166,7 @@ public class VersionCommand extends BukkitCommand {
                 return;
             }
             versionWaiters.add(sender);
-            sender.sendMessage("Checking version, please wait...");
+            // PandaSpigot - silenced "Checking version, please wait..."
             if (!versionTaskStarted) {
                 versionTaskStarted = true;
                 new Thread(new Runnable() {
@@ -185,8 +190,33 @@ public class VersionCommand extends BukkitCommand {
     private void obtainVersion() {
         String version = Bukkit.getVersion();
         if (version == null) version = "Custom";
+        // PandaSpigot start
+        if (version.startsWith("git-PandaSpigot-") || version.startsWith("git-EnoSpigot-")) {
+            String[] parts = version.substring(version.indexOf("-", 4) + 1).split("[-\\s]");
+            int distance;
+            try {
+                distance = fetchGitDistance("mimiflo/EnoSpigot", "master", parts.length > 1 ? parts[1] : parts[0]);
+            } catch (NumberFormatException e) {
+                distance = -1;
+            }
+
+            switch (distance) {
+                case -1:
+                    this.setVersionMessage("\u00A78\u2503 \u00A77Statut     \u00A78 \u00bb \u00A7eErreur de v\u00e9rification");
+                    break;
+                case 0:
+                    this.setVersionMessage("\u00A78\u2503 \u00A77Statut     \u00A78 \u00bb \u00A7a\u00c0 jour");
+                    break;
+                case -2:
+                    this.setVersionMessage("\u00A78\u2503 \u00A77Statut     \u00A78 \u00bb \u00A7eVersion inconnue");
+                    break;
+                default:
+                    this.setVersionMessage("\u00A78\u2503 \u00A77Statut     \u00A78 \u00bb \u00A7c" + distance + " version(s) de retard\n\u00A78\u2503 \u00A77T\u00e9l\u00e9charger\u00A78 \u00bb \u00A76github.com/mimiflo/EnoSpigot");
+                    break;
+            }
         // PaperSpigot start
-        if (version.startsWith("git-PaperSpigot-")) {
+        } else if (version.startsWith("git-PaperSpigot-")) {
+        // PandaSpigot end
             String[] parts = version.substring("git-PaperSpigot-".length()).split("[-\\s]");
             int paperSpigotVersions = getDistance("paperspigot", parts[0]);
             if (paperSpigotVersions == -1) {
@@ -246,6 +276,65 @@ public class VersionCommand extends BukkitCommand {
         }
     }
 
+    // PandaSpigot start
+    private static int fetchDownloadApiDistance(int runNumber) {
+        try {
+            HttpURLConnection connection = (HttpURLConnection) new URL("https://downloads.hpfxd.com/v2/projects/pandaspigot/versions/1.8.8").openConnection();
+            connection.setRequestProperty("Accept", "application/json");
+            connection.setRequestProperty("User-Agent", "EnoSpigot/" + Bukkit.getVersion());
+            connection.connect();
+            if (connection.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) return -2;
+            try (BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream(), Charsets.UTF_8))) {
+                JsonObject obj = new Gson().fromJson(reader, JsonObject.class);
+                JsonArray arr = obj.getAsJsonArray("builds");
+                int latest = StreamSupport.stream(arr.spliterator(), false)
+                    .mapToInt(JsonElement::getAsInt)
+                    .max()
+                    .orElse(-1);
+                if (latest < 0) {
+                    return -1;
+                } else {
+                    return latest - runNumber;
+                }
+            } catch (JsonSyntaxException | NumberFormatException e) {
+                e.printStackTrace();
+                return -1;
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+            return -1;
+        }
+    }
+    private static int fetchGitDistance(String repo, String branch, String hash) {
+        hash = hash.replace("\"", "");
+        try {
+            HttpURLConnection connection = (HttpURLConnection) new URL("https://api.github.com/repos/" + repo + "/compare/" + branch + "..." + hash).openConnection();
+            connection.setRequestProperty("Accept", "application/vnd.github.v3+json");
+            connection.setRequestProperty("User-Agent", "EnoSpigot/" + Bukkit.getVersion());
+            connection.connect();
+            if (connection.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) return -2; // Unknown commit
+            try (BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream(), Charsets.UTF_8))) {
+                JsonObject obj = new Gson().fromJson(reader, JsonObject.class);
+                String status = obj.get("status").getAsString();
+                switch (status) {
+                    case "identical":
+                        return 0;
+                    case "behind":
+                        return obj.get("behind_by").getAsInt();
+                    default:
+                        return -1;
+                }
+            } catch (JsonSyntaxException | NumberFormatException e) {
+                e.printStackTrace();
+                return -1;
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+            return -1;
+        }
+    }
+    // PandaSpigot end
+
     private static int getDistance(String repo, String currentVerInt) { // PaperSpigot
         try {
             BufferedReader reader = Resources.asCharSource(
